iree_tests/conftest.py

```diff
@@ -328,6 +338,13 @@ def test_run(self):
         proc = subprocess.run(self.run_args, capture_output=True, cwd=self.test_cwd)
         if proc.returncode != 0:
             raise IreeRunException(proc, self.test_cwd, self.compile_args)
+        
+    def test_benchmark(self):
+        proc = subprocess.run(self.benchmark_args, capture_output=True, cwd=self.test_cwd)
+        if proc.returncode != 0:
+            raise IreeRunException(proc, self.test_cwd, self.compile_args)
+        outs = proc.stdout.decode("utf-8")
+        print(f"Stdout benchmark:\n{outs}\n")
```

Comment:
Before this sort of change lands, let's think a bit about what we actually want coverage for. I'm skeptical about having benchmarks built into the same testing flow... though the suite that Stella set up has these:
* https://github.com/openxla/iree/blob/2c88e49d184b621c763a5cfb4af693f8dcbc6a07/experimental/regression_suite/ireers/fixtures.py#L81-L94
* https://github.com/openxla/iree/blob/2c88e49d184b621c763a5cfb4af693f8dcbc6a07/experimental/regression_suite/tests/pregenerated/test_llama2.py#L128-L148

Reply:
Just because we care so much about sdxl perf, I think it would be great to have it included in this flow. I didn't look into adding a whole separate flow for it or making it very scalable because I doubt we will be adding benchmarks for anything else. Probably just hardcoded commands for the backends we want to test perf on. I was thinking we can evaluate and iterate if this becomes a bigger utility but for now, just went for easiest/simple add for benchmarking. Here is an example log with everything running: https://github.com/nod-ai/SHARK-TestSuite/actions/runs/8543035287/job/23405926540

Reply:
The value of continuous benchmarks is clear, but I want to be careful about how we integrate them. For this PR, can you leave the benchmarks off and focus on just adding the sdxl models? A follow-up PR can then add benchmarking.

I'd at least like to take some time to work through the specific requirements before jumping straight to an implementation. For example:
* What metrics/artifacts do we want from benchmarking?
  * Each model in isolation? Full pipeline latency? Just dispatch time?
* What do we want done with benchmark results / artifacts?
  * The in-tree benchmarks in IREE submit results to a dashboard (that should use a queryable database...), upload Tracy files to cloud storage, and comment on pending pull requests with results summaries
* Where do we want benchmarks to run?
  * Right after tests, on presubmit to IREE?
  * In a separate job, on separate runners?

I'm also wondering if we want to use pytest as the benchmark runner (either with the existing conftest.py or a forked one), or if we would want to use another runner (could start with a pile of scripts, just using the same test suite source files)

Reply:
It might be reasonable to start with `pytest -k benchmark` or `pytest iree_tests/benchmarks` that just run `iree-benchmark-module` instead of `iree-run-module` and then let developers dig through the GitHub Actions logs to see results, but I'm worried about going down the path of building an entirely new benchmark "framework" when we already have https://github.com/openxla/iree-comparative-benchmark and https://github.com/openxla/iree/tree/main/build_tools/benchmarks (building something new is likely going to make sense, at least in the short term, but this stuff gets complicated very quickly)

Reply:
> this stuff gets complicated very quickly

For example: https://github.com/openxla/community/blob/main/rfcs/20230505-benchmarking-strategy.md

Reply:
Yeah, I definitely understand where you're coming from. I also thought it would be cool to have all the sdxl stuff in one place (model accuracy and benchmarking results) for developers to easily see results. At least for now, I am treating this more of a corner case. I'm all onboard to change it down the line if we feel the need to.

Reply:
At a minimum I'd put the benchmark changes in their own pull request. Then I'd also have review comments like:

* `.github/workflows/test_iree.yml` could be forked to `.github/workflows/benchmark_iree.yml` (particularly for the `nodai-amdgpu-w7900-x86-64` matrix runner config change)
* `iree_tests/conftest.py` should not hardcode any paths. We can add new config (JSON, txt, etc.) files as needed and have a programmatic branch on whether to construct a "test" test case and/or a "benchmark" test case

Reply:
> > this stuff gets complicated very quickly
> 
> For example: https://github.com/openxla/community/blob/main/rfcs/20230505-benchmarking-strategy.md

Wow, this is a very methodical, detailed way to approach it :0

Reply:
> At a minimum I'd put the benchmark changes in their own pull request. Then I'd also have review comments like:
> 
> * `.github/workflows/test_iree.yml` could be forked to `.github/workflows/benchmark_iree.yml` (particularly for the `nodai-amdgpu-w7900-x86-64` matrix runner config change)
> * `iree_tests/conftest.py` should not hardcode any paths. We can add new config (JSON, txt, etc.) files as needed and have a programmatic branch on whether to construct a "test" test case and/or a "benchmark" test case

Makes sense, will create another PR for it

Reply:
Also the runner change is not just for benchmarking. Just one of the sdxl submodels takes a hr to run on the github runners we were using before. It was this exact job that made me make the switch lol: https://github.com/nod-ai/SHARK-TestSuite/actions/runs/8505671886

Reply:
Ah... I'd rather not switch to the GPU machine because CPU tests are slow. That might be what we have as runners today, but we should be a bit more principled about it.

We can split the tests across runners:
* ONNX tests on a standard GitHub-hosted CPU machine
* Full model tests on a large self-hosted CPU (or GPU) machine, possibly split even further by model

Reply:
Ah yeah, that's a good idea. I will also make ONNX testing and model testing different jobs then (think only way to split by runner is having them on different jobs).

Reply:
> (think only way to split by runner is having them on different jobs).

I believe that can be done with a matrix too: https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#example-using-a-multi-dimension-matrix

Could then check other matrix parameters to choose which steps to run... maybe like this:

```yml
jobs:
  iree_tests:
    strategy:
      matrix:
        configuration:
          - runner: ubuntu-22.04
            test_command: "pytest iree_tests/simple --durations=0"
          - runner: ubuntu-22.04
            test_command: "pytest iree_tests/onnx/node/generated -n auto -rpfE --timeout=30 --retries 2 --retry-delay 5 --durations=10"
    runs-on: ${{ matrix.runner }}
    steps:
      ...
```

I'm also referencing https://github.com/openxla/iree/blob/573ff1ff02347266ed747dd316cefaeb4c710396/.github/workflows/ci.yml#L749-L784 (probably tons of other files to reference across GitHub, but that's what I know already...)

Reply:
If we wanted to plug in to the in-tree benchmark infrastructure that IREE has, we'd want a PR like https://github.com/openxla/iree/pull/16965 . That would feed into https://perf.iree.dev/ and PR comments, but it doesn't also test correctness out of the box, can be tricky to update (multiple Python files, coupled with GitHub Actions), and puts input files / parameters behind a few levels of abstraction that make it harder to run locally.

Reply:
Yeah, I also wonder how that would work because we need to essentially compile multiple submodels and then use their vmfbs for the pipeline. Not sure if it is setup for a pipeline structure.

Reply:
Also, I've placed the onnx and model tests in different jobs. I think that's best for this suite. Because they don't depend on each other and are running independently on different machines, I don't think we need the sequential steps. This way we have parallel execution which can help for scalability in future. Once we get more machines, splitting on models also would be great :)

Reply:
Here is the PR for benchmarking that should be landed after this one: https://github.com/nod-ai/SHARK-TestSuite/pull/155. Feel free to add on notes there for future reference